### PR TITLE
:package: Use `nodeenv` when Node.js can't be found by Python package

### PR DESCRIPTION
**Pull request by agoose77**
Fixes https://github.com/jupyter-book/mystmd/issues/1443 by using `nodeenv` to provision a Node.js environment for installs of our Python package.

Here's an example running in Podman (the final crash appears container related)
https://github.com/user-attachments/assets/9ca90520-f1fa-4d37-a6da-a64edc5f0a83

### CODE REVIEW

File: packages/mystmd-py/.gitignore [hunk 1]
```diff
@@ -3,3 +3,5 @@ dist
 src/mystmd_py/myst.cjs
 **/__pycache__/
 _package.json
+
+/_build/
```
Comment (rowanc1) [inline packages/mystmd-py/.gitignore:7]: Probably you running myst in this folder?!

File: packages/mystmd-py/src/mystmd_py/main.py [hunk 2]
```diff
@@ -4,41 +4,87 @@
...
+
+def find_any_node(binary_path):
+    node_path = find_installed_node()
+    if node_path is not None:
+        return pathlib.Path(node_path).absolute(), binary_path
+
+    nodeenv_path = find_nodeenv_path()
+    if not nodeenv_path.exists():
+        print(f"🔍 Couldn't find installed `node`.\n\n⚙️ Installing Node.js in {nodeenv_path}")
+        create_nodeenv(nodeenv_path)
```
Comment (rowanc1) [inline packages/mystmd-py/src/mystmd_py/main.py:45]: We should give the user control here - "Would you like to install node now using nodeenv?" y/n

File: packages/mystmd-py/src/mystmd_py/main.py [hunk 3]
```diff
@@ -4,41 +4,87 @@
...
     if not (major_version in {18, 20, 22} or major_version > 22):
         raise SystemExit(
             f"MyST requires node 18, 20, or 22+; you are running node {major_version}.\n\n"
             "Please update to the latest LTS release, using your preferred package manager\n"
             "or following instructions here: https://nodejs.org/en/download"
         )
+
+    # Start MyST CLI
+    PATH_TO_BIN_JS = (pathlib.Path(__file__).parent / "myst.cjs").resolve()
     os.execve(
```
Comment (rowanc1) [inline packages/mystmd-py/src/mystmd_py/main.py:84]: Somewhat unrelated: There is a bug on windows (maybe others) where this is execution does not get keyboard commands (so the y/n input stuff doesn't work at all). Do you have any idea how to fix that @agoose77?

File: packages/mystmd-py/.gitignore (hunk 1 above)
Comment (agoose77) [reply packages/mystmd-py/.gitignore:7]: Lol the one time I just use git add update!

File: packages/mystmd-py/src/mystmd_py/main.py [hunk 4]
```diff
@@ -4,41 +4,99 @@
...
+    return platformdirs.user_data_path(
+        appname="myst", appauthor=False, version=NODEENV_VERSION
     )
-    PATH_TO_BIN_JS = (pathlib.Path(__file__).parent / "myst.cjs").resolve()
 
-    if not NODE_LOCATION:
+
+def ask_to_install_node(path):
+    if env_value := os.environ.get(INSTALL_NODEENV_KEY, "").lower():
+        return env_value in {"yes", "true", "1"}
```
Comment (agoose77) [inline packages/mystmd-py/src/mystmd_py/main.py:26]:
```suggestion
        return env_value in {"yes", "true", "1", "y"}
```

**Comment by agoose77** [comment]
@rowanc1 I've added an `input` call (that can be bypassed by defining the `MYSTMD_ALLOW_NODENV` env var, where the result depends upon the truthiness (`{"yes", "true", "1", "y"}`) of the lowercased value.

File: packages/mystmd-py/src/mystmd_py/main.py (hunk 3 above)
Comment (agoose77) [reply packages/mystmd-py/src/mystmd_py/main.py:96]: I can't immediately see why the exec isn't working properly. will need to play around on Windows later.

**Comment by rowanc1** [comment]
Can you rebase this on the docs PR that just got merged. My understanding is that this will simplify the pip install instructions?

File: packages/mystmd-py/.gitignore (hunk 1 above)
Comment (rowanc1) [inline packages/mystmd-py/.gitignore:7]:
```suggestion
```

File: packages/mystmd-py/pyproject.toml [hunk 5]
```diff
@@ -20,6 +20,10 @@ classifiers = [
     "Programming Language :: Python :: 3.12",
     "Topic :: Scientific/Engineering",
 ]
+dependencies = [
+  "platformdirs~=4.2.2",
+  "nodeenv~=1.9.1"
```
Comment (rowanc1) [inline packages/mystmd-py/pyproject.toml:25]: How does this play with conda? Do we have other node deps over there that do the install for us?

File: packages/mystmd-py/.gitignore (hunk 1 above)
Comment (agoose77) [reply packages/mystmd-py/.gitignore:7]: Lol yep, still todo.

File: packages/mystmd-py/pyproject.toml (hunk 5 above)
Comment (agoose77) [reply packages/mystmd-py/pyproject.toml:25]: Oh right, that's a good point. I didn't think about how we would repackage that. Let me have a think.

File: packages/mystmd-py/pyproject.toml [hunk 6]
```diff
@@ -34,6 +34,14 @@ myst = "mystmd_py.main:main"
...
 path = "_package.json"		
 
+# Ensure that we only bring in nodeenv for PyPI
+[tool.hatch.build.hooks.custom]
+path = "build_plugin.py"
+pypi_dependencies = [
+  "platformdirs~=4.2.2",
+  "nodeenv~=1.9.1"
+]
+
```
Comment (agoose77) [inline packages/mystmd-py/pyproject.toml:44]: @rowanc1 this is a build plugin for hatch that injects the PyPI dependencies at build time. The conda-forge build pulls in the sdist (which is mainly just a tar of the package contents), so it does not have this injected content. When the conda-forge build happens, we'll need to update the environment to set MYST_IS_CONDA_PACKAGE, which will disable the injection.

**Comment by agoose77** [comment]
I have lifted the custom build hook into https://github.com/jupyter-book/hatch-deps-selector/, which means that others can also use it.

Simply put, this plugin makes it possible for us to add additional predefined dependencies to our package metadata at build-time by setting an environment variable.

**Comment by agoose77** [comment]
This PR now should hold until https://github.com/conda-forge/staged-recipes/pull/27417 is merged, so that we can immediately update our feedstock recipe.

**Comment by agoose77** [comment]
Adding a pre-built wheel to test on Windows...
[mystmd-1.3.6-py3-none-any.whl.zip](https://github.com/user-attachments/files/16847273/mystmd-1.3.6-py3-none-any.whl.zip)

**Comment by agoose77** [comment]
@rowanc1 this has been tested on Windows and Linux, and works as expected. We can review during today's meeting if helpful.

**Pull request merged**